napari/layers/image/_image_slice.py

```diff
@@ -0,0 +1,51 @@
+from typing import Optional
+
+from ._image_view import ImageView
+from ...types import ArrayLike, ImageConverter
+
+
+class ImageSlice:
+    """The slice of the image that we are currently viewing.
+
+    Right now this just holds the image and its thumbnail, however future async
+    and multiscale-async changes will likely grow this class a lot.
+
+    Parameters
+    ----------
+    view_image : ArrayLike
+        The default image for the time and its thumbail.
+    image_converter : ImageConverter, optional
```

Comment:
I think in the docstring (rather than the actual type annotation) I would prefer this to be `Callable, mapping array->array`, or something in that vein. When I read this, I see a crazy new object. Nothing in this docstring actually tells me, as a reader and longtime user of the scipy ecosystem, that `lambda x: x % 256` is an ImageConverter!

Reply:
The description below should indeed explain what imageconverter should be, rather than how imageslice uses it, since readers typically read this docstring to figure out what they need to pass / can pass this constructor.

Reply:
I agree in a way it would be nice to see the underlying type is `Callable[[ArrayLike], ArrayLike]` here. But I think that raises two questions:

1. Do we really want the type hint and doc string to purposely conflict? To me that looks like a mistake, and I see that mistake everywhere. Having the same type duplicated in two places where one is a comment is super error prone, against DRY/SPOT. They will inevitably get out of sync. But we are stuck with that format until they update things to suck out the type hints for docs.

2 If we keep them in sync and put `Callable[[ArrayLike], ArrayLike]` in both places are we just saying user defined types are not allowed? If not then when are they allowed or event encouraged? We have to spell that out I think, otherwise it's personal preference and thus hard to review.

In my experience using types without user defined types is the worst of all worlds and is a rookie mistake for people not used to typed languages. That said, I agree it can be overdone. Maybe some project has an existing style guide with some well thought out advice? I didn't see one googling for a bit.

Here's an example of where not using user defined types gets you:

<img width="599" alt="beazley-types" src="https://user-images.githubusercontent.com/4163446/84517070-bc2a8400-ac9c-11ea-8607-65414031471f.png">

Reply:
> Do we really want the type hint and doc string to purposely conflict?

I've never seen the numpy docstrings as very useful for actually typing things.  There is no standard, and they're nearly impossible to parse.  So it's just there as a human readable string (and any type information is also just a human readable suggestion).  There _are_ ways to pull out [type hints for sphinx docs](https://pypi.org/project/sphinx-autodoc-typehints/)... so technically we don't need the type hints in the numpy docstrings at all.  But they are helpful as a suggestion if someone is looking at docstrings in an interactive session... I just don't see them as binding in any way, or something that can really be "synced" successfully with the (actual) type annotation.

I'm not sure what you mean by the second point.  If there's a standard type that we use in lots of places (like `ReaderFunction`, or `LayerData`), then we should definitely use type aliases like that.  But fwiw, i disagree with that twitter post.  I _like_ having those annotations like that.  It's immediately self-documenting, and I don't have to go searching around for what it means.  (And also, if you _really_ hate that... you can put the type hints into a separate `.pyi` file)

Reply:
I can see arguments on both sides, and like the idea of finding a well respected style guide we can conform too (if such a thing exists!).

Given that this is something that we can very easily change and we'll be revisiting this part of the code shortly I'd like to merge now so as not hold up development (I'll be working on the world coordinates later today and so would like this in). This definitely doesn't mean though that I want to stop this convo or consider it settled.

Reply:
I've heard this complaint many times about types, and about "too many" functions and methods. People feel like stuff is being "hidden from them" and they just want to see it all at once. Totally a guess but I feel like maybe 20% of programmers have this complaint. I could give you a list of names from over the years.

To me the twitter example above is beyond horrific. I wouldn't include code like that in any project I was working on unless my specific intent was to hinder the development process and sink that project. So that gives? How can reasonable people have 180 degree out of phase opinions on stuff like this? Over the years I've concluded two things:

1. People just think differently. Some people like to see all the parts spread all over the table and other people don't. There's no way or reason to try to change either person, they are just wired that way. FWIW I feel like the "smarter" people tend to like to see all the details at once, because they can make sense of it. And crucially they simply can't imagine how anyone would not like to see it that way.

2. It's possible to mentally or physically "undo" user defined types, or too many small functions or methods, with tools or by hand, and create the "all details at once view". But it's totally impossible to do the reverse. A good type system, or functional decomposition, reflects a **huge** amount of cognitive work, carefully breaking complex things down and coming up with meaningful names. So between the two approaches I think it's clear the one with more information and more work behind it has to win out.

Still this is one of those things where people will never agree, but there needs to be project style where people say "I personally would not write it like that, but it fits our style". I'm not asserting what the right style is for napari, but it needs to develop one or reviews will always go in circles as people push for their personal style.

Finally I can't remember where I first read this, but this is a really important anecdote. People like to say lines of code doesn't matter, but it matters a lot by order of magnitude. Consider a 200 line program vs. 2000, 20,000, 200,000, 2M, 20M, 200M. I've worked on 200 to 2M but not bigger.

What this person was saying, which I totally agree with, is if you are used to one size and you look at the next larger size you don't think "wow that's great code" you think "What the heck are they doing? That looks really clunky. Why are they making it so complicated? That's crazy I'd *never* write it like that. I could do it *much* more simply."

But if you expand your 2,000 line program into a 20,000 line program without adopting the styles of the next size up, everything grounds to halt and your codebase becomes an usable mess. Suddenly you are like "oh.... now I see why they did it that way".

You see this "size discrepancy" all the time online because people argue about code without revealing what type of project they are working on. It's like you have 3 builders heatedly talking over lunch about building techniques, but one is building a dog house, one is building an big box store, and one is building a nuclear power plant. So they are all right but also all wrong.

Reply:
I'll attempt a TLDR:
* Programmers sometimes have *very* different ideas about what style of code is easier or harder to understand or work with. Sometimes the exact opposite instincts.
* Projects need a good clear style so reviews can focus on that and not personal preferences.
* Some styles that work well in smaller codebases fall apart in larger ones.
* Corollary: some styles that work well in large codebases can be clunky in smaller codebases.

Also I was thinking that for me the case for using a type aliases or user defined type goes up rapidly the more times that type appears in the code. Maybe there is a trade-off point where the benefits of "factoring out the duplication" outweigh the burden of "one more level of indirection".

Reply:
> Also I was thinking that for me the case for using a type aliase or user defined type gets stronger the more times that a type appears in the code.

that's what I meant by pointing out `LayerData` and `ReaderFunction`, where we do just that.

Reply:
> Also I was thinking that for me the case for using a type alias or user defined type gets stronger the more times that a type appears in the code.

That's exactly right, and I see a short future for ImageConverter, especially in a world in which NumPy is starting to develop its own annotation stubs, so an array -> array function will probably get its own name from there.

But, this is a bit of a side discussion on the main one, which was the NumPy style docstring type annotations. To answer your earlier questions, @pwiston,

- I don't think it's a huge problem if they diverge from the actual type annotation. One annotation is meant for humans, another for machines. Attempts to join the two modes have usually failed.
- the text description of the argument in the docstring certainly has *no* restrictions and that should clarify what we are talking about here.
- I actually completely disagree with Talley that these annotations are not useful — this is literally the first place I check when exploring a new API and if the object types are simple, the description is always sufficient for me to infer what I need to do. Contrast this with the standard library, which doesn't have this convention and I am often left wondering what exactly needs to go in that argument.

And of course, going back to the ImageConverter type alias — for this specific case, it is so simple that indeed I think `Callable[[np.ndarray], np.ndarray]` is preferable. It's not even that many more characters. And in fact, I think that type is too precise, since a callable mapping a dask array to a dask array would work here also. So it could just be `Callable`, or `Callable[[ArrayLike], ArrayLike]`.

> Maybe some existing Python project's style guide discusses this?

I think type hints are too new in Python for any style guides to have made pronouncements on this. But in Dave Beazley's example, I don't think type aliases would help. He would still hate the signature and it would be even less clear, unless, as discussed above, the type aliases are extremely common or totally self-explanatory. `ImageConverter` is neither of those, nor do I foresee it becoming so.

Reply:
> I actually completely disagree with Talley that these annotations are not useful

? I feel like we're saying the exact same thing!  so I think you misunderstood my point.  I didn't say they are not useful "full stop"... I said they are not useful for actually typing things. (in a way that can be programmatically parsed), and I said they are there as human readable... exactly where a human might go first when exploring a new API, or in an interactive session.

Reply:
Oops! Sorry for misunderstanding your point! Agreement is good! 😅

Reply:
This PR is merged, can this change be part of #1354? If the answer to when to use type aliases is "sometimes" we should flesh out in more detail what the guidelines are, when to use which style. Long term IDE's should let you toggle showing expanded types. You can see expanded types on hove in vscode. That seems to be heading towards the best of both worlds.

<img width="569" alt="ArrayLike" src="https://user-images.githubusercontent.com/4163446/84572894-20b41480-ad6b-11ea-9c33-6b508e4eaf48.png">

<img width="813" alt="LayerData" src="https://user-images.githubusercontent.com/4163446/84572898-2578c880-ad6b-11ea-8311-4e02da8c9858.png">

Reply:
Those screenshots are pretty nice @pwinston. Maybe we continue this convo on #1355 and then if we do make changes have them be as part of a dedicated PR. I'd rather avoid mixing it in with #1354 where we'll have lots to discuss too.

Reply:
My personal position on IDEs is that they are nice for people that have them, but policies should not depend on people using an IDE. Code should just be readable. Probably 90% of my interaction with code these days is reading it on GitHub PRs, so until GitHub itself becomes an IDE, plaintext interpretation is the lowest common denominator that we should target.